### PR TITLE
Automatic mapping from quantization method to quantizer

### PR DESCRIPTION
**Pull request by ofirgo**
Replacing the manual mapping from quantization target and method to quantizer in quantization infrastructure using a new `@mark_quantizer` decorator.
The main changes include:
- Implementing the `@mark_quantizer` decorator, which should be used to decorate all quantizers class (currently, QAT and Inferable quantizers). It defines static properties that allow to classify the quantizer - quantization methodm quantization target and quantizer type.
- Adding a middle layer for QAT trainable quantizers - BaseKerasQATTrainableQuantizer and BasePytorchQATTrainableQuantizer.
- Removing the global mapping dictionaries from node_to_quantizer and quantizer_builder and replacing them with an automatic selection of the appropriate quantizer.
- Removing QuantizationMethod import from QI.
- Updating Readmes for adding a new quantizer.
- Other minor cosmetic changes.

### CODE REVIEW

File: model_compression_toolkit/quantizers_infrastructure/pytorch/README.md [hunk 1]
```diff
@@ -2,29 +2,39 @@
...
+[`BasePytorchTrainableQuantizer`]([https://github.com/sony/model_optimization/blob/main/model_compression_toolkit/quantizers_infrastructure/keras/base_Keras_quantizer.py](https://github.com/sony/model_optimization/blob/main/model_compression_toolkit/quantizers_infrastructure/pytorch/base_pytorch_quantizer.py) constitutes a base class for trainable quantizers of specific of specific tasks--currently, [`BasePytorchQATTrainableQuantizer`](https://github.com/ofirgo/model_optimization/blob/301a6c024828075bf1e6429de6a2750f26732170/model_compression_toolkit/qat/pytorch/quantizer/base_pytorch_qat_quantizer.py) for Quantization-Aware Training.
+
+### The mark_quantizer Decorator
+The [`mark_quantizer`](https://github.com/ofirgo/model_optimization/blob/301a6c024828075bf1e6429de6a2750f26732170/model_compression_toolkit/quantizers_infrastructure/common/base_inferable_quantizer.py) decorator is used to supply each quantizer with static properties which define its task compitability. Each quantizer class should be decorated with this decorator. It defines the following properties:
```
Comment (elad-c) [inline model_compression_toolkit/quantizers_infrastructure/pytorch/README.md]: this is your fork
relative links will solve this

File: model_compression_toolkit/quantizers_infrastructure/pytorch/base_pytorch_quantizer.py [hunk 2]
```diff
@@ -12,41 +12,34 @@
...
 
 else:
     class BasePytorchTrainableQuantizer(BaseTrainableQuantizer):
         def __init__(self,
-                     quantization_config: Union[TrainableQuantizerWeightsConfig, TrainableQuantizerActivationConfig],
-                     quantization_target: QuantizationTarget,
-                     quantization_method: List[QuantizationMethod]):
-            super().__init__(quantization_config, quantization_target, quantization_method)
+                     quantization_config: Union[TrainableQuantizerWeightsConfig, TrainableQuantizerActivationConfig]):
+            super().__init__(quantization_config)
```
Comment (elad-c) [inline model_compression_toolkit/quantizers_infrastructure/pytorch/base_pytorch_quantizer.py:42]: BasePytorchTrainableQuantizer

**Review by elad-c — CHANGES_REQUESTED** [review]

File: model_compression_toolkit/qat/pytorch/quantizer/quantization_builder.py [hunk 3]
```diff
@@ -12,34 +12,60 @@
...
+    get_trainable_quantizer_quantization_candidates, get_trainable_quantizer_weights_config, \
+    get_trainable_quantizer_activation_config
+from model_compression_toolkit.qat.pytorch.quantizer.base_pytorch_qat_quantizer import BasePytorchQATTrainableQuantizer
+from model_compression_toolkit.quantizers_infrastructure import QuantizationTarget
+from model_compression_toolkit.quantizers_infrastructure.common.constants import QUANTIZATION_TARGET, \
+    QUANTIZATION_METHOD, QUANTIZER_TYPE
+from model_compression_toolkit.quantizers_infrastructure.common.get_all_subclasses import get_all_subclasses
+
+
+def _get_quantizer_class(quant_target: QuantizationTarget,
```
[the remark this answers is not in the document]
Comment (ofirgo) [reply model_compression_toolkit/qat/pytorch/quantizer/quantization_builder.py:32]: It's only common in qat and in qi (for each of them there still needs to be implemented a separate _get_quantizer_class).
I'll move this to common in each of the packages, but it wouldn't eliminate all the duplication.

File: model_compression_toolkit/exporter/model_wrapper/keras/builder/node_to_quantizer.py [hunk 4]
```diff
@@ -106,6 +84,36 @@ def get_inferable_quantizer_kwargs(node: BaseNode,
...
 
+def _get_quantizer_class(quant_target: QuantizationTarget,
+                         quant_method: QuantizationMethod) -> type:
+    """
+    Searches for a quantizer class that matches the requested QuantizationTarget and QuantizationMethod.
+    Exactly one class should be found.
+
+    Args:
+        quant_target: QuantizationTarget value which indicates what is the target for quantization to
+            use the quantizer for.
```
Comment (Idan-BenAmi) [inline model_compression_toolkit/exporter/model_wrapper/keras/builder/node_to_quantizer.py]: maybe you can add for simplicity: weights or activation

File: model_compression_toolkit/quantizers_infrastructure/common/base_trainable_quantizer.py [hunk 5]
```diff
@@ -17,52 +17,60 @@
...
         self.quantization_config = quantization_config
-        self.quantization_method = quantization_method
-        if self.quantization_target == QuantizationTarget.Weights:
+
+        # Inherited class should be decorated with @mark_quantizer decorator, and define the following static properties
+        static_quantization_method = getattr(self, QUANTIZATION_METHOD)
+        static_quantization_target = getattr(self, QUANTIZATION_TARGET)
+
+        if static_quantization_method is None or static_quantization_target is None:
+            # TODO: add test that defines a new quantizer without the decorator and check for the exception.
```
[the remark this answers is not in the document]
Comment (ofirgo) [reply model_compression_toolkit/quantizers_infrastructure/common/base_trainable_quantizer.py]: Added.

File: model_compression_toolkit/quantizers_infrastructure/pytorch/README.md [hunk 6]
```diff
@@ -1,82 +1,15 @@
 ## Introduction
 
-[`BasePytorchTrainableQuantizer`](https://github.com/sony/model_optimization/blob/main/model_compression_toolkit/quantizers_infrastructure/pytorch/base_pytorch_quantizer.py) is an interface that enables easy quantizers development and training. 
+[`BasePytorchTrainableQuantizer`](./base_pytorch_quantizer.py) is an interface that enables easy quantizers development and training. 
 Using this base class makes it simple to implement new quantizers for training and inference for weights or activations.
+[`BasePytorchTrainableQuantizer`](./base_pytorch_quantizer.py) constitutes a base class for trainable quantizers of specific of specific tasks - currently, [`BasePytorchQATTrainableQuantizer`](../../qat/pytorch/quantizer/base_pytorch_qat_quantizer.py) for Quantization-Aware Training.
```
Comment (Idan-BenAmi) [inline model_compression_toolkit/quantizers_infrastructure/pytorch/README.md:5]: trainable quantizers of specific **of specific** tasks

File: model_compression_toolkit/exporter/model_wrapper/common/__init__.py [hunk 7]
```diff
@@ -0,0 +1,14 @@
+# Copyright 2022 Sony Semiconductor Israel, Inc. All rights reserved.
```
Comment (haihabi) [inline model_compression_toolkit/exporter/model_wrapper/common/__init__.py:1]: 2023

File: model_compression_toolkit/exporter/model_wrapper/pytorch/builder/node_to_quantizer.py [hunk 8]
```diff
@@ -90,6 +70,36 @@ def get_activation_inferable_quantizer_kwargs(node: BaseNode) -> Dict[str, Any]:
         Logger.critical(f'Not supported quantization method for inferable quantizers.')  # pragma: no cover
 
 
+def _get_quantizer_class(quant_target: QuantizationTarget,
```
[the remark this answers is not in the document]
Comment (ofirgo) [reply model_compression_toolkit/exporter/model_wrapper/pytorch/builder/node_to_quantizer.py]: Yes I moved it to a common location

**Review by Idan-BenAmi — APPROVED** [review]

**Pull request merged**